code/payments/Tally-Vs-Rollup.md

```diff
@@ -0,0 +1,69 @@
+# Tally vs Rollup
+
+The terms "tally" and "rollup" are used throughout the code base. Their specific meanings have been a source of confusion since their meanings are similar.
+
+The purpose of this document is to:
+- define the terms "tally" and "rollup".
+- provide details about how tally and rollup are used in the code base.
+
+## Definitions
+
+Both tally and rollup sum an amount of something (i.e. storage or bandwidth) as it relates to a period of time.
+
+They differ in what the period of time is.
+- For tally, the period of time is just a single moment in time, typically the most recent or current moment.
+- For rollup, the period of time is a time frame, for example 1 hr or 1 day time interval.
+
+This means that tally is a snapshot of the current state. And rollup is an aggregate of those snapshots.
+
+## Implementations of Tally and Rollup
+
+Here is a high level overview of how tallies and rollups are used in the code base.
+
+The items that need to be tallied and rolledup include:
+- bandwidth usage for storage nodes
+- bandwidth usage for uplinks
+- storage usage for storage nodes
+- storage usage for uplinks
+
+#### Bandwidth usage is tallied when Orders are created
+
+Details:
+
+OrderLimits are created by the Satellite (SA) when an Uplink uploads/downloads a file. OrderLimits indicate the maximum amount of bandwidth an Uplink can use per Storage node (SN) when storing a file. Orders are created by the Uplink when storing a file on a SN.  The Uplink sends the Orders to the SN. The Order has an "amount" field which is the "tally" and indicates how much bandwidth the Uplink is currently using. The SN saves this and once the file is done being uploaded, then sends the Orders to the SA to finanlize (i.e. "settle") that amount.
```

Comment:
`Amount` is not the tally, it's the exact amount that is allowed to be paid to a specific storage node.

Reply:
But the `amount` is what gets aggregated in the rollup, so isn't the `amount` playing the role of tallying? Even though obviously the amount is was is to be paid.

Reply:
You can rollup things that are not tallies. (e.g. rollup of a rollup, rollup of values)

Reply:
I'm trying to simplify the concept of tally and rollup so that people can understand them easier at first.  Of course when trying to simplify something that is complex, one of the trade offs is accuracy.  

What I'm trying to show here is that when thinking of "tally" and "rollup", the simplest way to understand them and their relationship to each other is that tally counts things in the current moment, and rollup aggregates those tally counts over a period of time.

If we extend this concept to the places in the codebase, we need to think about where we tally and where we rollup storage and bandwidth usage.   So the question is: where does the tally happen and where does the rollup happen for both storage and bandwidth usage?

In the current implementation, the amount on the `Order` is where the "tally" is occurring for the bandwidth usage.  Then those "tallies" are rolled-up when the `Order` is settled on the SA.

Yes you can rollup things that are not tallies, but for simplicity Im only trying to describe the main use cases currently in the code base.

Reply:
I do think it may actually end up being clearer to only assume storage is tallied, and bandwidth is not tallied at all. We don't have tallied counts of storage amounts without tallying them, but the bandwidth agreement does already have settled bandwidth amounts. 

Perhaps it would be a simpler line to draw that "tallies" are really only something that the Tally service does, and rollups can apply to any value?

Reply:
I do see the line you're drawing though and it's not wrong at all. That said, if we define tally to be a point-in-time value, then I think a lot more things end up being able to be called "tallies" then we really maybe mean here. It might not be the best scalpel

Reply:
Ultimately the goal of tallying and rolling up is to invoice storage and bandwidth usage.  I mostly want to describe the workflow for those.  If we can draw parallels between the storage and bandwidth invoicing workflows I think it helps to make more sense of it. 

However if in doing that actually makes it less clear, then I need to rework this doc and not treat the accounting of bandwidth workflow as similar steps as the accounting of storage workflow.